lib/interface_quda.cpp

```diff
@@ -3252,346 +3230,389 @@ void invertQuda(void *hp_x, void *hp_b, QudaInvertParam *param)
-  // std::vector<ColorSpinorField*> x;  // Cuda Solutions
-  // x.resize(param->num_src);
-  ColorSpinorField* in;  // = nullptr;
-  //in.resize(param->num_src);
-  ColorSpinorField* out;  // = nullptr;
-  //out.resize(param->num_src);
+  gauge_param->ga_pad = fat_pad;
+  if (inv_param->dslash_type == QUDA_STAGGERED_DSLASH || inv_param->dslash_type == QUDA_LAPLACE_DSLASH) {
+    gauge_param->reconstruct = link_recon;
+    gauge_param->reconstruct_sloppy = link_recon_sloppy;
+    gauge_param->reconstruct_refinement_sloppy = link_recon_sloppy;
+  } else {
+    gauge_param->reconstruct = QUDA_RECONSTRUCT_NO;
+    gauge_param->reconstruct_sloppy = QUDA_RECONSTRUCT_NO;
+    gauge_param->reconstruct_refinement_sloppy = QUDA_RECONSTRUCT_NO;
+  }
+  gauge_param->reconstruct_precondition = QUDA_RECONSTRUCT_NO;
 
-  // for(int i=0;i < param->num_src;i++){
-  //   in[i] = nullptr;
-  //   out[i] = nullptr;
-  // }
+  loadGaugeQuda(milc_fatlinks, gauge_param);
 
-  const int *X = cudaGauge->X();
+  if (inv_param->dslash_type == QUDA_ASQTAD_DSLASH) {
+    gauge_param->type = QUDA_ASQTAD_LONG_LINKS;
+    gauge_param->ga_pad = link_pad;
+    gauge_param->staggered_phase_type = QUDA_STAGGERED_PHASE_NO;
+    gauge_param->reconstruct = link_recon;
+    gauge_param->reconstruct_sloppy = link_recon_sloppy;
+    gauge_param->reconstruct_refinement_sloppy = link_recon_sloppy;
+    gauge_param->reconstruct_precondition = link_recon_precondition;
+    loadGaugeQuda(milc_longlinks, gauge_param);
+  }
+}
 
+template <class Interface, class... Args>
+void callMultiSrcQuda(void **_hp_x, void **_hp_b, QudaInvertParam *param, // color spinor field pointers, and inv_param
+                      void *h_gauge, void *milc_fatlinks, void *milc_longlinks,
+                      QudaGaugeParam *gauge_param,     // gauge field pointers
+                      void *h_clover, void *h_clovinv, // clover field pointers
+                      Interface op, Args... args)
+{
+  /**
+    Here we first re-distribute gauge, color spinor, and clover field to sub-partitions, then call either invertQuda or dslashQuda.
+    - For clover and gauge field, we re-distribute the host clover side fields, restore them after.
+    - For color spinor field, we re-distribute the host side source fields, and re-collect the host side solution fields.
+  */
 
-  // Host pointers for x, take a copy of the input host pointers
-  void** hp_x;
-  hp_x = new void* [ param->num_src ];
+  profilerStart(__func__);
 
-  void** hp_b;
-  hp_b = new void* [param->num_src];
+  const int *_split_key = param->split_grid;
+  CommKey split_key = {_split_key[0], _split_key[1], _split_key[2], _split_key[3]};
+  int num_sub_partition = quda::product(split_key);
 
-  for(int i=0;i < param->num_src;i++){
-    hp_x[i] = _hp_x[i];
-    hp_b[i] = _hp_b[i];
+  for (int d = 0; d < CommKey::n_dim; d++) {
+    if (split_key[d] < 1) { errorQuda("split_key[%d] = %d is non-positive.\n", d, split_key[d]); }
   }
```

Comment:
Should this be a member of Comm_Key? Like Comm_Key:is_valid() ? 
And why do we need a cast to `int*` for split key? Can't we use the values from param->split_grid directly?